borsh/tests/test_recursive_structs.rs

```diff
@@ -0,0 +1,66 @@
+#![cfg(feature = "derive")]
+use borsh::{from_slice, BorshDeserialize, BorshSerialize};
+
+#[cfg(feature = "hashbrown")]
+use hashbrown::HashMap;
+
+#[cfg(feature = "std")]
+use std::collections::HashMap;
+
+#[cfg(hash_collections)]
+use core::{cmp::Eq, hash::Hash};
+
+#[cfg(not(feature = "std"))]
+extern crate alloc;
+#[cfg(not(feature = "std"))]
+use alloc::{boxed::Box, string::String, vec::Vec};
+
+#[cfg(hash_collections)]
+#[derive(BorshSerialize, BorshDeserialize)]
+struct CRec<U: PartialOrd + Hash + Eq> {
+    a: String,
+    b: HashMap<U, CRec<U>>,
+}
```

Comment:
strangely enough, this worked before current pr. (at least compiled)
Somehow added complexity of generic param `U` broke rust compiler out of recursion